dot/state/tries.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package state
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ChainSafe/chaindb"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/trie"
+)
+
+type tries struct {
+	rootToTrie map[common.Hash]*trie.Trie
+	db         chaindb.Database
+	mapMutex   sync.RWMutex
+}
+
+func newRootToTrieMap(db chaindb.Database, t *trie.Trie) *tries {
+	return &tries{
+		rootToTrie: map[common.Hash]*trie.Trie{
+			t.MustHash(): t,
+		},
+		db: db,
+	}
+}
+
+// getValue retrieves the value in the trie specified by trieRoot
+// and at the node with the key given. It returns an error if
+// the value is not found.
+// Note it does NOT cache the trie in memory if it is not found
+// in memory but found in the database, unlike the getTrie method.
+func (t *tries) getValue(trieRoot common.Hash, key []byte) (
+	value []byte, err error) {
+	// Try to get from memory
+	t.mapMutex.RLock()
+	tr, has := t.rootToTrie[trieRoot]
+	t.mapMutex.RUnlock()
+	if has {
+		value = tr.Get(key)
+		// TODO: should we return an error if the value is not found?
+		return value, nil
```

Comment:
Question here! ❓ 

When a key is not found in the database, we return an error. However, from memory we don't, should we change that? Or return a `nil` value from the database if not found?

Reply:
@qdm12 I was looking at the [GetFromDB](https://github.com/ChainSafe/gossamer/blob/6c64521f57f24556489cb2183bf19b1e2f02ab79/lib/trie/database.go#L280) and [getFromDB](https://github.com/ChainSafe/gossamer/blob/6c64521f57f24556489cb2183bf19b1e2f02ab79/lib/trie/database.go#L306), the functions whose retrieves the data from database, and the returned errors means problems while getting data (eg. database closed or database doesn't exists) or decoding the data (eg. incomplete or wrong encoded hashes), however, if we can retrieve and decode the hashes and even though the key is not present both functions will return nil without error, so I think we don't need to change the behavior. Please, correct me if I'm wrong

Reply:
Awesome thanks! Not sure why I thought it was returning an error when not found...

I also moved changes regarding atomic operations in #2250 so I'll remove that `TODO`